server/server.ts

```diff
@@ -0,0 +1,372 @@
+          } catch (err) {
+            showErrorMessage(docPath, `Unable to read ruleset at ${rulesetFile}.`);
+          }
+        } else {
+          // If there's no ruleset available, default to OAS rules.
+          // https://github.com/stoplightio/spectral/blob/4928b6155659cec225d0cfab6d77d849d691cc24/src/cli/services/linter/utils/getRuleset.ts#L22
+          settings.ruleset = await readRuleset('spectral:oas');
+        }
+      }
+
+      return settings;
+    });
+
+  documentSettings.set(uri, resultPromise);
+  return resultPromise;
+}
+
+/**
+ * Registers for document events and starts listening for those events.
+ */
+function setupDocumentListener(): void {
+  documents.listen(connection);
+
+  documents.onDidOpen((event) => {
+    resolveSettings(event.document).then((settings) => {
+      if (!settings.validate || settings.run !== 'onSave') {
+        return;
+      }
+      messageQueue.addNotificationMessage(ValidateNotification.type, event.document, event.document.version);
+    });
+  });
+
+  documents.onDidChangeContent((event) => {
+    resolveSettings(event.document).then((settings) => {
+      if (!settings.validate || settings.run !== 'onType') {
+        return;
+      }
+      messageQueue.addNotificationMessage(ValidateNotification.type, event.document, event.document.version);
+    });
+  });
+
+  documents.onDidSave((event) => {
+    resolveSettings(event.document).then((settings) => {
+      if (!settings.validate || settings.run !== 'onSave') {
+        return;
+      }
+      messageQueue.addNotificationMessage(ValidateNotification.type, event.document, event.document.version);
+    });
+  });
+
+  documents.onDidClose((event) => {
+    resolveSettings(event.document).then((settings) => {
+      const uri = event.document.uri;
+      documentSettings.delete(uri);
+      if (settings.validate) {
+        // Clear any diagnostics associated with the document that closed.
+        connection.sendDiagnostics({ uri: uri, diagnostics: [] });
+      }
+    });
+  });
+}
+
+/**
+ * Pops up toast to say something bad happened, directs the user to a logged console message.
+ * @param {string} uri The document being validated when the error occurred.
+ * @param {any} err The error that occurred.
+ */
+function showErrorMessage(uri: string, err: any): void {
+  connection.window.showErrorMessage(`Spectral: An error occurred while validating ${uri}. Please see the 'Spectral' output channel for details.`);
+  connection.console.error(`An error occurred while validating document ${uri}: ${err}`);
+}
+
+/**
+ * Validates a single document with Spectral and presents the results to VS Code as diagnostics.
+ * @param {TextDocument} document - The document to validate.
+ * @return {Thenable<void>} A promise to await completion.
+ */
+function validate(document: TextDocument): Thenable<void> {
+  return resolveSettings(document).then(async (settings) => {
+    if (!settings.validate) {
```

Comment:
@tillig Could you please help me understand the conditions under which this would happen?

Reply:
@tillig Scratch that. I got it.